rse-r-package.Rmd

```diff
@@ -0,0 +1,590 @@
+
+Once you start writing code for other people---including yourself three months from now---your
+documentation should describe:
+
+1.  The name and purpose of every function and constant in your code.
+2.  The name, purpose, and default value (if any) of every parameter to every function.
+3.  Any side effects the function has.
+4.  The type of value returned by every function.
+5.  Why and how the function will deliberately fail.
+    If a function uses something like `stopifnot` or `assert` to check that a condition holds,
+    then that halt-and-catch-fire behavior is effectively part of its interface.
+
+### Exercise: Fixing documentation
+
+FIXME: provide poorly-documented function and ask learner to find and fix gaps.
+
+### Exercise: Cross-references
+
+1.  What should you add to the roxygen2 comments for one function to link to the documentation for another function?
+2.  Add a cross-reference from the documentation for the function FIXME in the CO2 package to the documentation for the function FIXME.
+
+### Exercise: Documenting error conditions
+
+The guidelines above said that authors should document why and how their functions will deliberately fail.
+Where and how should you do this using roxygen2 for R?
+
+## How do I manage package dependencies? {#rse-r-package-dependencies}
+
+Let's add another function in another file.
+(We don't really need the function,
+and it doesn't have to go in another file,
+but we'll create it for tutorial purposes.)
+
+```{r, child="rse-r-package/etc/counter-original.R"}
+```
+
+This function depends on functions from the tidyverse---more specifically,
+from the dplyr package.
+Since our compiled-and-distributable package will only contain the compiled code for its own functions,
+direct calls to functions from other libraries won't work after the package is installed.
+To fix this,
+we add this to the roxygen2 comment block for our function:
+
+```{r eval=FALSE}
+#' @import dplyr
+#' @importFrom magrittr %>%
+```
+
+to tell the package builder that we want all of dplyr available,
+plus the `%>%` operator from magrittr.
+We then various calls to use their package prefix:
+
+```{r eval=FALSE}
+tally_words <- function(all_words) {
+  dplyr::tibble(word = all_words) %>%
+    dplyr::group_by(word) %>%
+    dplyr::tally()
+}
+```
+
+Running a check gives us the following error:
+
+```
+❯ checking R code for possible problems ... NOTE
+  tally_words: no visible global function definition for ‘%>%’
+  tally_words: no visible global function definition for ‘tibble’
+  tally_words: no visible global function definition for ‘group_by’
+  tally_words: no visible binding for global variable ‘word’
+  tally_words: no visible global function definition for ‘tally’
+  Undefined global functions or variables:
+    %>% group_by tally tibble word
+```
+
+What we need to do is add these three lines to the bottom of `DESCRIPTION`:
+
+```text
+Imports:
+    dplyr (>= 0.7.0),
+    magrittr (>= 1.5.0)
```

Comment:
Better to use:

```
usethis::use_package("dplyr")
```

Reply:
Catching my angle here? :wink: usethis is really great for package development! :grin:

Reply:
Yes, but we still need to describe what goes into the file so that they can debug when things go wrong.